### PR TITLE
fix(config): split multi-pattern entries with unique installation IDs per org

### LINKED ISSUE #34: [Bug] Multi-pattern entries with single installation ID create invalid configuration
### Describe the bug

When configuring multiple GitHub Apps with different organization patterns in a single configuration entry, the installation ID is computed during setup and remains static. This causes the "test" mode to fail because it cannot correctly validate credentials against multiple organizations with a single installation ID, which is logically invalid.

The issue arises when a configuration entry contains multiple patterns (e.g., `github.com/gh-app-auth-01`, `github.com/gh-app-auth-02`) but a single installation ID. In reality, each organization should have its own installation ID, making this configuration state invalid.

### Affected version

0.0.6

### Steps to reproduce the behavior

1. Export the private key environment variable:
   ```bash
   export GH_APP_PRIVATE_KEY
   ```

2. Run the setup command with multiple patterns:
   ```bash
   gh app-auth setup \
     --app-id 2892328 \
     --patterns "github.com/gh-app-auth-01" \
     --patterns "github.com/gh-app-auth-02"
   ```

3. Observe that a single installation ID is assigned to both patterns in the resulting configuration
4. Attempt to run `gh app-auth test` - it will fail because the installation ID is not valid for both organizations

### Expected vs actual behavior

**Expected behavior:** When multiple patterns are provided during setup, the system should either:
- Reject the configuration as invalid (single installation ID cannot serve multiple organizations), OR
- Automatically split the configuration into separate entries, one per organization pattern

**Actual behavior:** The system creates a single configuration entry with multiple patterns and one installation ID, which is logically invalid. The "test" mode fails when attempting to validate credentials against all patterns using the same installation ID.

### Configuration

Example of problematic configuration that results from the bug:

```yaml
version: "1.0"
github_apps:
    - name: GitHub App 2892328
      app_id: 2892328
      installation_id: 110896322  # Single ID for multiple orgs - INVALID!
      private_key_source: keyring
      patterns:
        - github.com/gh-app-auth-01
        - github.com/gh-app-auth-02
      priority: 5
```

- GitHub App ID: 2892328
- Patterns configured: Multiple patterns (github.com/gh-app-auth-01, github.com/gh-app-auth-02)
- Repository URL: Any repository under the configured organizations

### Environment

- OS: Linux
- GitHub CLI version: `gh version`
- Go version (if building from source): `go version`

### Logs

Run with debug flag for verbose logs:
```
gh-app-auth --debug setup --app-id 2892328 --patterns "github.com/gh-app-auth-01" --patterns "github.com/gh-app-auth-02"
```

### Root Cause

The configuration system allows multiple patterns per entry but computes a single installation ID during setup. This creates a logical inconsistency: a single installation ID cannot be valid for multiple organizations simultaneously. The "test" mode fails because it attempts to validate against all patterns using the same credentials.

**Note:** The actual git commands behave correctly because they dynamically compute all parameters on-the-fly for each repository, avoiding the static installation ID problem.

### Proposed Solution

When processing configuration entries with multiple patterns:

1. Detect entries with multiple organization patterns and a single installation ID
2. Split these entries into separate configuration entries, one per organization pattern
3. Ensure each pattern gets its own installation ID entry
4. Prevent the creation of logically invalid configuration states where one installation ID maps to multiple organizations

This ensures that:
- Each organization has its own dedicated installation ID
- The "test" mode can correctly validate credentials
- Configuration remains logically consistent

### Additional context

- This is a logical consistency issue that affects configuration validation
- The actual git operations work correctly because they compute parameters dynamically
- The "test" mode and configuration validation fail due to the static installation ID mismatch
- Users may encounter this when setting up multiple organizations with a single GitHub App

### PR DESCRIPTION
**Pull request by wherka-ama**
## Description

Fixes a bug where `gh app-auth setup` incorrectly handles multiple organization patterns by assigning a single installation ID to all patterns, regardless of which organizations they reference. Since each GitHub App installation is scoped to a single organization, this leads to authentication failures when using the "test" mode or when accessing repositories from different organizations.

When users run:
```bash
gh app-auth setup --app-id 28xx \
  --patterns "github.com/xx" \
  --patterns "github.com/yy"
```

The previous implementation would create a single config entry with one installation ID (from the first org) applied to all patterns. This fix ensures each organization gets its own correctly detected installation ID in separate config entries.

Fixes #34

## Type of Change

Please delete options that are not relevant:

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔧 Refactoring (no functional changes, no api changes)
- [ ] ⚡ Performance improvements
- [ ] 🧪 Test updates
- [ ] 🔒 Security improvements

## Changes Made

- **Added `extractUniqueOrgs()` function**: Parses patterns and extracts host/organization information for each pattern
- **Added `groupPatternsByOrg()` function**: Groups patterns by unique organization to identify how many separate config entries are needed
- **Added `validateMultiPatternSetup()` function**: Validates multi-org pattern formats before processing
- **Modified `setupGitHubApp()` function**: Core fix that iterates over unique organizations, auto-detects installation ID per org, and creates separate `GitHubApp` config entries for each organization
- **Added `displaySetupSuccessMultiOrg()` function**: Displays success message showing multiple organizations configured with their respective patterns
- **Added comprehensive tests**: New test functions covering all scenarios (single org, multiple orgs, enterprise hosts, edge cases) with >85% coverage

## Testing

Please describe the tests that you ran to verify your changes:

- [x] Unit tests pass: `go test ./...`
- [x] Build succeeds: `go build -o gh-app-auth .`
- [x] Manual testing performed
- [ ] Integration tests with real GitHub Apps (if applicable)

### Test Configuration

- **OS**: Linux
- **Go version**: 1.23+
- **GitHub CLI version**: 2.67+

**New Test Coverage**:
- `TestExtractUniqueOrgs` - 9 test cases covering single org, multiple orgs, enterprise hosts, empty/invalid patterns
- `TestGroupPatternsByOrg` - 3 test cases for pattern grouping logic
- `TestSetupMultiOrgErrors` - 4 test cases for validation error handling
- `TestDisplaySetupSuccessMultiOrg` - 3 test cases for multi-org success message display

All tests pass with >85% coverage on new functions:
- `extractUniqueOrgs`: 100%
- `validateMultiPatternSetup`: 100%
- `displaySetupSuccessMultiOrg`: 100%
- `groupPatternsByOrg`: 85.7%

## Security Considerations

If this PR involves security-related changes:

- [x] No sensitive data (tokens, keys) exposed in code or tests
- [x] Input validation added/reviewed for new functionality  
- [x] Private key handling follows security best practices
- [x] Token caching security maintained
- [x] No new attack vectors introduced

## Documentation

- [x] Code comments updated
- [ ] README.md updated (if needed)
- [ ] Architecture documentation updated (if needed)
- [ ] Command help text updated (if needed)

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published
- [x] **Commit messages follow [Conventional Commits](https://github.com/AmadeusITGroup/gh-app-auth/blob/main/CONTRIBUTING.md#commit-message-guidelines) format**
- [x] **PR title follows conventional commits format** (e.g., `feat(auth): add token caching`)

## Breaking Changes

If this is a breaking change, please describe the impact and migration path for existing users:

**No breaking changes.** This is a bug fix that changes internal behavior to produce correct output. The CLI interface remains unchanged - users still use the same commands. The fix only affects the internal configuration generation logic to produce valid multi-org configurations instead of invalid single-entry configurations.

**For users with existing invalid configurations**: They should re-run `gh app-auth setup` with their patterns to generate correct split configurations. The old invalid single-entry configurations can be removed via `gh app-auth remove`.

### Root Cause Analysis
The bug was in `setupGitHubApp()` function in `cmd/setup.go`:
1. `autoDetectInstallationID()` only used the first pattern to detect installation ID
2. `createGitHubApp()` created a single entry with all patterns sharing that one ID
3. Result: Config with one installation ID mapped to multiple organizations (logically invalid)

### Solution Approach
- Group patterns by unique organization using `groupPatternsByOrg()`
- For each org, call `autoDetectInstallationID()` with a representative pattern
- Create separate `GitHubApp` entry per organization with its unique installation ID
- Handle edge cases: same org with multiple patterns = single entry, different orgs = separate entries

### Validation Performed
- All existing tests continue to pass
- New tests added for multi-org scenarios
- Coverage maintained at >50% (project requirement)
- Build successful with no new warnings
- No race conditions or flaky tests introduced

### Files Changed
- `cmd/setup.go` - Core implementation (modified)
- `cmd/setup_test.go` - Comprehensive tests (modified)

**Commit**: `4a5f88b` - `fix(config): split multi-pattern entries with unique installation IDs per org`

### CODE REVIEW

File: cmd/setup.go [hunk 1]
```diff
@@ -565,3 +656,37 @@ func displaySetupSuccess(
...
+	// Display storage information
+	if backend == secrets.StorageBackendKeyring {
+		fmt.Printf("   🔐 Storage: OS Keyring (encrypted)\n")
+		if expandedKeyFile != "" {
+			fmt.Printf("   📝 Fallback: %s\n", expandedKeyFile)
+		}
+	} else {
+		fmt.Printf("   📁 Storage: Filesystem\n")
+		fmt.Printf("   📝 Key file: %s\n", expandedKeyFile)
+		fmt.Printf("   ⚠️  Keyring unavailable - using filesystem storage\n")
```
Comment (ph--) [inline cmd/setup.go]: Unecessary `fmt.Printf`

**Review by ph-- — CHANGES_REQUESTED** [review]
